backend/uclapi/workspaces/occupeye/archive.py

```diff
@@ -0,0 +1,217 @@
+
+                # Sensors where the hardware id or survey device id have changed will have two sets of data
+                # We want to keep the data with the greatest hardware ID or survey device ID
+                unique_sensors = {}
+                for key, sensor in enumerate(sensors):
+                    unique_key = str(sensor["SensorID"]) + ":" + sensor["TriggerDate"]
+                    unique_value = sensor["HardwareID"] + sensor["SurveyDeviceID"]
+                    if unique_key in unique_sensors:
+                        if unique_value > unique_sensors[unique_key][1]:
+                            unique_sensors[unique_key] = (key, unique_value)
+                    else:
+                        unique_sensors[unique_key] = (key, unique_value)
+
+                filtered_sensors = []
+                for _, value in unique_sensors.items():
+                    filtered_sensors.append(sensors[value[0]])
+
+                sensors = filtered_sensors
+                batch_objects = []
+
+                for sensor in sensors:
+                    cs_key = str(sensor["SensorID"]) + f"-{survey.survey_id}"
+                    if cs_key in self.current_sensors:
+                        self.validate_sensor(sensor, survey, start_date, end_date)
+                    else:
+                        Sensors.objects.create(sensor_id=sensor["SensorID"], hardware_id=sensor["HardwareID"],
+                                               survey_device_id=sensor["SurveyDeviceID"],
+                                               survey_id=survey.survey_id)
+                        self.current_sensors[str(sensor["SensorID"]) + "-" + str(survey.survey_id)] = {
+                            "hardware_id": sensor["HardwareID"], "survey_device_id": sensor["SurveyDeviceID"],
+                            "survey_id": survey.survey_id, "last_value": -1}
+
+                    for key, value in sensor.items():
+                        if key.startswith("t"):
+                            # A few time stamps have None instead of -1
+                            if value is None:
+                                value = -1
+                            if value != self.current_sensors[cs_key]["last_value"]:
+                                updates += 1
+                                time = datetime.strptime(sensor["TriggerDate"] + key, "%Y-%m-%dt%H%M")
+                                time = time.replace(tzinfo=timezone.utc)
+                                batch_objects.append(
+                                    Historical(sensor_id=sensor["SensorID"], survey_id=survey.survey_id,
+                                               datetime=time, state=value))
+                                self.current_sensors[cs_key]["last_value"] = value
+
+                # Ideally ignore conflicts would be false but depending on exactly when the archive is run or
+                # if the previous run crashed some conflicts may exist.
```

Comment:
They are effectively all in one transaction as the sensors one each survey are independent. This was to prevent the task crashing if we run it twice within one day or very close to midnight.

Reply:
Why would it crash if we ran it twice a day? This comes across as fragile, I'd like to avoid that if possible.

Reply:
Sorry I think that was badly phrased, the code **should** never crash.

If it is run twice in one day, the last_updated field shows that it has already run and so it will fetch a time period of 0 (thus not requesting anything). However I think there might be an edge case if it is run around midnight with the endpoint not yet having the data for the final hour of the day (although I am not entirely sure).

However in the event that the code does crash for some reason, for example the endpoint going down for longer than the period of time for 3 connection attempts then some of the survey locations may not be populated. If this happens halfway through updating one of the surveys then it will not insert the data into the database which has already been saved. If this occurs between survey locations then last_updated would have been changed and it will not be run for that survey.